src/mapgen.cpp

```diff
@@ -3571,17 +3571,53 @@ class jmapgen_nested : public jmapgen_piece
                     return true;
                 }
         };
+        // TO DO: Remove this and replace it with a smarter logical syntax for neighbor_join_check and neighbor_flag_check
+        class neighbor_flag_any_check
+        {
+            private:
+                std::unordered_map<direction, cata::flat_set<oter_flags>> neighbors;
+            public:
+                explicit neighbor_flag_any_check( const JsonObject &jsi ) {
+                    for( direction dir : all_enum_values<direction>() ) {
+                        cata::flat_set<oter_flags> dir_neighbours;
+                        std::string location = io::enum_to_string( dir );
+                        optional( jsi, false, location, dir_neighbours );
+                        if( !dir_neighbours.empty() ) {
+                            neighbors[dir] = std::move( dir_neighbours );
+                        }
+                    }
+                }
+
+                bool test( const mapgendata &dat ) const {
+                    bool any_direction_has_flag = false;
+                    for( const std::pair<const direction, cata::flat_set<oter_flags>> &p :
+                         neighbors ) {
+                        const direction dir = p.first;
+                        const cata::flat_set<oter_flags> &allowed_flags = p.second;
+
+                        cata_assert( !allowed_flags.empty() );
+
+                        for( const oter_flags &allowed_flag : allowed_flags ) {
+                            any_direction_has_flag |=
+                                dat.neighbor_at( dir )->has_flag( allowed_flag );
+                        }
+                    }
+                    return any_direction_has_flag;
```

Comment:
Optional: you could return true if `dat.neighbor_at( dir )->has_flag( allowed_flag )` to skip the rest of the nested for loop iterations.